### PR TITLE
Dev/rgroup decomp freefunction

### PR DESCRIPTION
**Pull request by bp-kelley**
Adds free function API and python interface.

This also fixes a subtle but serious bug in the RGroupDecomp.  The RGroupMatch data should have been stored as a shared pointer as early copies were not maintained properly between iterations.  The multicore results from the cpp tests were returning hydrogens that should have been stripped out.  This fixes that issue and assigns the rgroups indices properly (previously, the first four had incorrect assignments).

If this is acceptable, the docstrings are the last item necessary.

### CODE REVIEW

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.h [hunk 1]
```diff
@@ -143,6 +143,19 @@ class RGroupDecomposition {
   //! return rgroups in column order group[attachment_point][row] = ROMol
   RGroupColumns getRGroupsAsColumns() const;
 };
+
+unsigned int RGroupDecompose(RGroupRows &rows,
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupDecomp.h]: Minor thing: in the interest of consistency with the rest of the API (or at least most of it), the argument used to return results  - `rows` here, `cols` below - should be the last non-optional argument. So this function would be `RGroupDecompose(cores,mols,rows,unmatched,options)`

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.h [hunk 2]
```diff
@@ -143,6 +143,19 @@ class RGroupDecomposition {
...
+                             const std::vector<ROMOL_SPTR> &cores,
+                             const std::vector<ROMOL_SPTR> &mols,
+                             std::vector<unsigned int> *unmatched = 0,
+                             const RGroupDecompositionParameters &options = RGroupDecompositionParameters());
+
+unsigned int RGroupDecompose(RGroupColumns &rows,
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupDecomp.h]: `rows` should be `cols`
and see comment above about argument order

**Comment by greglandrum** [comment]
There are still two outstanding suggestions from my review. 
Would you rather keep the argument ordering as is or did those just overlooked?

**Comment by bp-kelley** [comment]
No, I missed those two, probably hyper focused when I found the bugs.

**Pull request merged**